bpf/protocol_http2.h

```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
     bpf_map_delete_elem(&ongoing_http2_grpc, stream);
 }
 
-static __always_inline void process_http2_grpc_frames(pid_connection_info_t *pid_conn,
-                                                      void *u_buf,
-                                                      int bytes_len,
-                                                      u8 direction,
-                                                      u8 ssl,
-                                                      u16 orig_dport) {
-    int pos = 0;
-    u8 found_start_frame = 0;
-    u8 found_end_frame = 0;
-
-    http2_grpc_request_t *prev_info = 0;
-    u32 saved_stream_id = 0;
-    int saved_buf_pos = 0;
-    u8 found_data_frame = 0;
-    http2_conn_stream_t stream = {0};
-    stream.pid_conn = *pid_conn;
-
-    unsigned char frame_buf[FRAME_HEADER_LEN];
-    frame_header_t frame = {0};
-
-    for (int i = 0; i < 8; i++) {
-        if (pos >= bytes_len) {
+static __always_inline frame_header_t next_frame(const grpc_frames_ctx_t *g_ctx) {
+    const call_protocol_args_t *args = &g_ctx->args;
+
+    frame_header_t header;
+
+    // read next frame
+    const void *offset = (u8 *)args->u_buf + g_ctx->pos;
+
+    if (bpf_probe_read(&header, sizeof(header), offset) != 0) {
+        bpf_dbg_printk("failed to read frame header");
+        goto header_error;
+    }
+
+    bpf_dbg_printk("http2 frame type = %d, len = %d", header.type, header.length);
```

Comment:
With regards to the `int` usage, my biggest concern is that we have a signed int here - in most APIs (except from  a few relics like libc functions such as read(), which will rely on negative values for errors, and anything else for "byte count"), we will encounter `size_t`, which tends to be the same as `uintptr_t` because it needs to be able to address the memory space. We do invariably end up mixing these ints up with pointers and alike, which are unsigned by default, so IMHO it's just best to eventually refactor them to become unsigned (and simplify underflow/overflow cases).

Using a smaller _unsigned_ int here is potentially fine, but we have to be careful to not make it appear as an lvalue in an operation involving pointers as rvalues, as it can cause an overflow (because pointers will be 64-bit). It becomes a trade-off between saving a few bytes (or a lot depending on how many of those we have) versus less cognitive load of treating everything as `size_t`.

Reply:
Done